File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int prod = 1;
        int prodWithoutZero = 1;
        int countNonZero = 0;
        for(auto ele: nums){
            prod *= ele;
            if(ele != 0){
                countNonZero++;
                prodWithoutZero *= ele;
                }
        }

        int zeroCount = nums.size() - countNonZero;
        vector<int> ans(nums.size(),0);
        if(zeroCount > 1)
            return ans;

        for(int i = 0; i < nums.size(); i++){
            if(nums[i] != 0)
                ans[i] = prod/nums[i];
            else
                ans[i] = prodWithoutZero;             
        }
        return ans;
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:

        prod, prod_wo, zeroCnt = 1, 1, 0

        for num in nums:
            if num == 0:
                zeroCnt += 1
                prod = 0
            else:
                prod *= num
                prod_wo *= num
        
        if zeroCnt > 1:
            return [0] * len(nums)
        
        # ans = []
        # for num in nums:
        #     if num == 0:
        #         ans.append(prod_wo)
        #     else:
        #         ans.append(prod//num)

        # return ans

        ans = [0] * len(nums)
        for i,num in enumerate(nums):
            if num == 0:
                ans[i] = prod_wo
            else:
                ans[i] = prod//num

        return ans


        